DENG-319

rebecca-burwei opened the pull request:
This fulfills [DENG-319](https://mozilla-hub.atlassian.net/browse/DENG-319).

I've completed steps 1-11 [here](https://mozilla.github.io/bigquery-etl/cookbooks/common_workflows/#adding-a-new-scheduled-query). The rest of the steps have to be done by someone with DE permissions.


Checklist for reviewer:

- [x] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title)
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](/CODEOWNERS)):
- [x] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

whd commented on dags/bqetl_search_terms_daily.py:
```diff
@@ -91,6 +91,24 @@
         arguments=["--schema_update_option=ALLOW_FIELD_ADDITION"],
     )
 
+    search_terms_derived__search_terms_daily__v1 = bigquery_etl_query(
```
See PR comment in `query.sql`. I don't think this can land as-is as airflow access to unsanitized search terms is very limited.

whd commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql:
```diff
@@ -0,0 +1,31 @@
+    DATE(timestamp) AS merino_date,
+    jsonPayload.fields.*
+  FROM
+    `suggest-searches-prod-a30f.logs.stdout`
```
Unsanitized search terms tables are not accessible to standard bqetl machinery: see the comment in https://github.com/mozilla/bigquery-etl/blob/main/sql/moz-fx-data-shared-prod/search_terms_derived/suggest_impression_sanitized_v2_external/query.sql#L3.

We currently have two production mechanisms for accessing unsanitized search terms data:
1. BQ scheduled query within the unsanitized project, as was done with `suggest_impression_sanitized_v2`
https://github.com/mozilla-services/cloudops-infra/blob/master/projects/merino/tf/modules/log-routing/sanitizer.sql
https://github.com/mozilla-services/cloudops-infra/blob/master/projects/merino/tf/modules/log-routing/main.tf#L187

2. via a custom k8s job https://github.com/mozilla/search-terms-sanitization

@chelseatroy may have an opinion on how to proceed here. The caveat for the scheduled query approach is that it requires a bqetl dag to copy the data from the merino project (where the scheduled query is contrained to write to) into shared-prod.

chelseatroy replied:
I _think_ the easiest thing might end up being a virtual view, really. We did this with the languages table: Some SQL runs on a regular cadence and populates that table right from bigquery-etl. That can be found in `bigquery-etl/sql/moz-fx-data-shared-prod/search_terms`. It consists of some metadata describing the required access and then a SQL query and that's it.

quiiver replied:
My understanding was that we were creating an aggregate of **sanitized** search terms.

whd replied:
If we're creating an aggregate of sanitized search terms then perhaps all that's needed here is to change the source dataset to that sanitized location and obviate the cross-project machinery required to access the unsanitized search terms.

rebecca-burwei replied:
Yes - we're just aggregating sanitized search terms. Sorry about the confusion! I fixed the SQL.

whd reviewed (commented):
I don't think this can land as-is; see PR comments.

This change also needs to follow the change control procedure (filing a DSRE ticket) linked at the bottom of the PR template, see https://mozilla-hub.atlassian.net/browse/DSRE-936 / https://mozilla-hub.atlassian.net/browse/DSRE-899 for recent tickets to use as reference. Since :mreid is out this week I imagine some combination of @quiiver and :mreid's manager should approve the change.

chelseatroy commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/metadata.yaml:
```diff
@@ -0,0 +1,19 @@
+friendly_name: Search Terms Daily
```
This sounds right to me.

rebecca-burwei replied:
updated

chelseatroy commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/metadata.yaml:
```diff
@@ -0,0 +1,19 @@
+description: |-
+  Daily count of terminal search terms collected by Merino.
+owners:
+- example@mozilla.com
```
I am happy to own this if that suits, and @rebecca-burwei we can also have multiple owners here if you'd like to put your email too!

rebecca-burwei replied:
Thanks @chelseatroy . I added both of us as owners

chelseatroy commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql:
```diff
@@ -0,0 +1,31 @@
+WITH query_logs AS (
```
Yiss. An example of that can be found in `/Users/chelseatroy/mozilla/bigquery-etl/sql/moz-fx-data-shared-prod/search_terms_derived/sanitization_job_metadata_v2/metadata.yaml`!

rebecca-burwei replied:
updated

rebecca-burwei commented:
I have initiated the change control procedure: https://mozilla-hub.atlassian.net/browse/DSRE-1029

rebecca-burwei commented:
Hi @whd -- I fixed the SQL to pull from sanitized search terms instead of unsanitized. I believe this should resolve most of the concerns. I've also added workgroup and owners to the metadata.yaml file. 

I also initiated the change control procedure: https://mozilla-hub.atlassian.net/browse/DSRE-1029

quiiver commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_weekly_v1/query.sql:
```diff
@@ -0,0 +1,22 @@
+  FROM
+    `moz-fx-data-shared-prod.search_terms_derived.merino_log_sanitized_v3` sanitized
+  WHERE
+    date(timestamp) = @submission_date
```
Is the goal to replace the weekly roll up on a daily basis? if so I think what we want here is something like:

```sql
WHERE
  date(timestamp) >= date_trunc(@submission_week, week)
```

rebecca-burwei replied:
@quiiver Thanks for the heads up. I was wrong about this -- we still want a daily table, but on each day, we'll look back at the past 7 days and only keep the search terms that have appeared >1 time in the last 7 days. Do you think [this](https://github.com/mozilla/bigquery-etl/blob/5e720044629dbfd27a29b813e8d35c4101bf0951/sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql#L7) and [this](https://github.com/mozilla/bigquery-etl/blob/5e720044629dbfd27a29b813e8d35c4101bf0951/sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql#L44) will work?

quiiver commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql:
```diff
@@ -0,0 +1,44 @@
+  FROM
+    `moz-fx-data-shared-prod.search_terms_derived.merino_log_sanitized_v3` sanitized
+  WHERE
+    DATE(timestamp) > @submission_date - 7
```
Its probably better to use `date_sub(@submission_date, interval 7 day)` here

quiiver commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql:
```diff
@@ -0,0 +1,44 @@
+        SUBMISSION_DATE ASC
+      ROWS BETWEEN
+        6 PRECEDING
+        AND CURRENT ROW
```
I don't think we need to bound the preceding window here if we're already setting a bound on the `terminal_queries` CTE. 

In fact, I think you can just drop the window clause here. from the docs:
> For aggregate analytic functions, if the ORDER BY clause is present but the window frame clause is not, the following window frame clause is used by default:
> 
> `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`

quiiver reviewed (approved):
looks good thanks, I think @whd will have to deploy this one

rebecca-burwei commented:
@whd we're waiting for final sign-offs [here](https://docs.google.com/document/d/1xPiQP-72B4MtgdIZRbyYJcwlKh5pA9yLZ8gxC6rrwEs/edit#heading=h.ddfi2mqlhlyw). After all the sign-offs are there, then we can make this data a reality.

rebecca-burwei commented:
@whd the sign offs are complete

whd commented:
@scholtzan  I'm attempting to deploy the table because I believe it won't be generated automatically by the DAG. `./bqetl query schema deploy` requires a `schema.yaml` file to exist, so I generated one via the tooling `schema update`:

```yaml
fields:
- mode: NULLABLE
  name: submission_date
  type: DATE
- mode: NULLABLE
  name: query
  type: STRING
- mode: NULLABLE
  name: search_sessions
  type: INTEGER
```

That file should perhaps be part of this PR. However, attempting to `schema deploy` I see the following error:

`google.api_core.exceptions.BadRequest: 400 POST https://bigquery.googleapis.com/bigquery/v2/projects/moz-fx-data-shared-prod/datasets/search_terms_derived/tables?prettyPrint=false: Clustering is enabled but no fields were specified.`

This looks like a "correct" error in that the results schema of `query.sql`  don't seem to include the partitioning field that is specified in `metadata.yaml`, namely `merino_date`. I'm presently of the opinion that the query needs to be changed as a result, but would appreciate you verifying that I haven't made any procedural errors here.

scholtzan commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/metadata.yaml:
```diff
@@ -0,0 +1,24 @@
+  incremental: true
+bigquery:
+  time_partitioning:
+    field: merino_date
```
Looking at the query, there is no `merino_date` field
```suggestion
    field: submission_date
```

scholtzan commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/metadata.yaml:
```diff
@@ -0,0 +1,24 @@
+    expiration_days: null
+  clustering:
+    fields: []
+references: {}
```
```suggestion
```
this isn't needed as it is later automatically added through generation

scholtzan commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/query.sql:
```diff
@@ -0,0 +1,41 @@
+        SUBMISSION_DATE ASC
+    ) AS search_sessions_in_last_7_days
+  FROM
+    DAILY_QUERIES
```
nit
```suggestion
    daily_queries
```

scholtzan commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/metadata.yaml:
```diff
@@ -0,0 +1,24 @@
+    type: day
+    require_partition_filter: true
+    expiration_days: null
+  clustering:
+    fields: []
```
```suggestion
```

scholtzan reviewed (commented):
There is no `merino_date` specified in the query, which should explain the error.

Ideally, the `schema.yaml` file should be added as part of this PR. The schema that is shown in https://github.com/mozilla/bigquery-etl/pull/3216#issuecomment-1317929752 looks correct.

whd commented:
@rebecca-burwei per Anna's feedback in https://github.com/mozilla/bigquery-etl/pull/3216#pullrequestreview-1184674267 can you add the `schema.yaml` file from https://github.com/mozilla/bigquery-etl/pull/3216#issuecomment-1317929752 ?

whd commented on sql/moz-fx-data-shared-prod/search_terms_derived/search_terms_daily_v1/metadata.yaml:
```diff
@@ -0,0 +1,31 @@
+scheduling:
+  dag_name: bqetl_search_terms_daily
+  arguments: ['--schema_update_option=ALLOW_FIELD_ADDITION']
+fields:
```
`schema.yaml` should its own file: https://mozilla.github.io/bigquery-etl/cookbooks/creating_a_derived_dataset/#creating-the-table-schema. It's needed to run step 13 of https://mozilla.github.io/bigquery-etl/cookbooks/common_workflows/.

whd reviewed (approved):
I was able to create the table. Once this is merged table ACLs will propagate and it will become accessible to users.
